packages/compiler/test/core/linter.test.ts

```diff
@@ -29,6 +34,65 @@ const noModelFoo = createLinterRule({
   },
 });
 
+const exitLintRuleSync = createLinterRule({
+  name: "exit-lint-rule-sync",
+  description: "",
+  severity: "warning",
+  messages: {
+    default: "Exit lint rule sync called",
+  },
+  async: false,
+  create(context: any) {
+    return {
+      interface: (target) => {
+        context.lastInterface = target;
+      },
+      exit: () => {
+        context.reportDiagnostic({
+          target: context.lastInterface,
+          messageId: "default",
+        });
+      },
+    };
+  },
+});
+
+const noInterfaceFooAsync = createLinterRule({
+  name: "no-interface-foo2-async",
+  description: "",
+  severity: "warning",
+  messages: {
+    default: "Cannot call interface 'Foo2' (async rule)",
+  },
+  async: true,
+  create(
+    context: LinterRuleContext<{
+      readonly default: "Cannot call interface 'Foo2' (async rule)";
+    }> & { interfaceToCheck?: Interface[] },
+  ) {
+    return {
+      interface: (target) => {
+        if (!context.interfaceToCheck) {
+          context.interfaceToCheck = [];
+        }
+        context.interfaceToCheck.push(target);
+      },
+      exit: async () => {
+        const r = await new Promise<Interface[]>((resolve) => {
+          setTimeout(() => {
+            resolve(context.interfaceToCheck?.filter((t) => t.name === "Foo2") ?? []);
+          }, 50);
```

Comment:
don't add time based timeout to test this just slows down for no reason. if it you do 0 it will already yield and behave async